app/src/org/commcare/connect/network/ApiConnectId.java

```diff
@@ -0,0 +1,400 @@
+                    expiration.setTime(expiration.getTime() + ((long)seconds * 1000));
+
+                    String seatedAppId = CommCareApplication.instance().getCurrentApp().getUniqueId();
+                    ConnectDatabaseHelper.storeHqToken(context, seatedAppId, hqUsername, token, expiration);
+
+                    return new AuthInfo.TokenAuth(token);
+                }
+            } catch (IOException | JSONException e) {
+                Logger.exception("Parsing return from HQ OIDC call", e);
+            }
+        }
+
+        return null;
+    }
+
+    public static ConnectNetworkHelper.PostResult makeHeartbeatRequestSync(Context context) {
+        String url = context.getString(R.string.ConnectHeartbeatURL);
+        HashMap<String, String> params = new HashMap<>();
+        String token = FirebaseMessagingUtil.getFCMToken();
+        if(token != null) {
+            params.put("fcm_token", token);
+            boolean useFormEncoding = true;
+            return ConnectNetworkHelper.postSync(context, url, API_VERSION_CONNECT_ID, retrieveConnectIdTokenSync(context), params, useFormEncoding, true);
+        }
+
+        return new ConnectNetworkHelper.PostResult(-1, null, null);
+    }
+
+    public static AuthInfo.TokenAuth retrieveConnectIdTokenSync(Context context) {
+        ConnectUserRecord user = ConnectDatabaseHelper.getUser(context);
+
+        if (user != null) {
+            AuthInfo.TokenAuth connectToken = user.getConnectToken();
+            if (connectToken != null) {
+                return connectToken;
+            }
+
+            HashMap<String, String> params = new HashMap<>();
+            params.put("client_id", "zqFUtAAMrxmjnC1Ji74KAa6ZpY1mZly0J0PlalIa");
+            params.put("scope", "openid");
+            params.put("grant_type", "password");
+            params.put("username", user.getUserId());
+            params.put("password", user.getPassword());
+
+            String url = context.getString(R.string.ConnectTokenURL);
+
+            ConnectNetworkHelper.PostResult postResult = ConnectNetworkHelper.postSync(context, url,
+                    API_VERSION_CONNECT_ID, new AuthInfo.NoAuth(), params, true, false);
+            if (postResult.responseCode == 200) {
+                try {
+                    String responseAsString = new String(StreamsUtil.inputStreamToByteArray(
+                            postResult.responseStream));
+                    postResult.responseStream.close();
+                    JSONObject json = new JSONObject(responseAsString);
+                    String key = ConnectConstants.CONNECT_KEY_TOKEN;
+                    if (json.has(key)) {
+                        String token = json.getString(key);
+                        Date expiration = new Date();
+                        key = ConnectConstants.CONNECT_KEY_EXPIRES;
+                        int seconds = json.has(key) ? json.getInt(key) : 0;
```

Comment:
I abstracted the code to a new SsoToken class and updated code to pass it around when convenient (instead of the token and expiration date independently). [Commit](https://github.com/dimagi/commcare-android/pull/2847/commits/b4dad7645467d93874873fc73347f7861b1e172f)